src/store-api/src/metadata.rs

```diff
@@ -694,9 +697,33 @@ impl RegionMetadataBuilder {
 
         for column_meta in self.column_metadatas.iter_mut() {
             if let Some(target_type) = change_type_map.remove(&column_meta.column_schema.name) {
-                column_meta.column_schema.data_type = target_type;
+                column_meta.column_schema.data_type = target_type.clone();
+                // also cast default value to target_type if default value exist
+                let new_default =
+                    if let Some(default_value) = column_meta.column_schema.default_constraint() {
+                        Some(
+                            default_value
+                                .cast_to_datatype(&target_type)
+                                .with_context(|_| CastDefaultValueSnafu {
+                                    default_value: Some(default_value.clone()),
+                                    target_type: target_type.clone(),
+                                })?,
+                        )
+                    } else {
+                        None
+                    };
+                column_meta.column_schema = column_meta
+                    .column_schema
+                    .clone()
+                    .with_default_constraint(new_default)
+                    .with_context(|_| CastDefaultValueSnafu {
+                        default_value: column_meta.column_schema.default_constraint().cloned(),
+                        target_type,
+                    })?;
```

Comment:
nitpick: Maybe we can change the CastDefaultValue to keep a message. So you can provide a different message to distinguish errors from `cast_to_datatype` and `with_default_constraint`.
```
CastDefaultValueSnafu {
    reason: format!("error message"),
}
```